Ts refactor tools

TomSmithCGAT opened the pull request:
@IanSudbery . Could you check over this before we merge into master.  I've removed the CGAT dependencies and moved the relevant code into Utilities.py. Setup.py now performs dependency checks based on the packages listed in required.txt. 

Cheers

IanSudbery commented on umi_tools/dedup.py:
```diff
@@ -553,7 +554,7 @@ def get_bundles(insam, ignore_umi=False, subset=None, quality_threshold=0,
                 do_output = not read.tid == last_chr
             else:
-                do_output = start > (last_pos+1000) or not read.tid == last_chr
+                do_output = start > (last_pos+1000) and not read.tid == last_chr
```
This is a regression. Make sure that the `or` version is in the final code when it is merged.

TomSmithCGAT replied:
Apologies for this. I had a similar issue with the changes you made to counting outputted reads - 968ee3be238263dcf0285a58b419ccf57f977614. I'm not sure how this happens given the branch was made from master after these commits?

IanSudbery commented:
Couple of things from just looking: 

1. By pull Experiment.py from CGAT, we have now a whole load of options that are totally irrelevant to the script.

2. We've always have a bit of a issue with the fact that logging and output are both sent to stdout, soif you pipe directly to an outfile, it won't be a valid BAM unless you redirect the log. Perhaps this is an oppotunity to do something about this (maybe we could tweak Experiment.py to sending logging to stderr?)

3. There is a regression on line 557(?) of dedup.py - i've comment in the code, the `and` should be an `or` - make sure the correct version is merged into master. 

I'll try to install and check it goes okay.

IanSudbery commented:
Ha... too slow

TomSmithCGAT commented:
1. Ah, I see what you mean about the unnecessary options being added with Utilities.Start(), e.g add_cluster_options. These are defaulted to False but the code is not needed so I'll remove these options.

2. I think directing the Utilities.py logging to stderr is a good idea for dedup.py but this should be optional. Perhaps add an options to Utilities.py ("log2stder") to determine whether to log to stdout or stderr.

TomSmithCGAT commented:
@IanSudbery I've removed the additional options in Utilities.Start() which we weren't using and added a common option "--log2stderr" which redirects logging to stderr if no log file has been specified.

TomSmithCGAT commented:
@IanSudbery Re: the cluster topologies. 3e24dc8e3d6837b4cdd0b1ec4ebc9468316c6a5d fixes the assignment. The previous code was a hangover from the adjacency dictionary function in place at the time.

IanSudbery commented:
I got the following error when I tried to setup:

```python-traceback
[ians@cgat150 UMI-tools]$ python setup.py  develop
running develop
running egg_info
creating umi_tools.egg-info
writing pbr to umi_tools.egg-info/pbr.json
writing requirements to umi_tools.egg-info/requires.txt
writing umi_tools.egg-info/PKG-INFO
writing top-level names to umi_tools.egg-info/top_level.txt
writing dependency_links to umi_tools.egg-info/dependency_links.txt
writing entry points to umi_tools.egg-info/entry_points.txt
writing manifest file 'umi_tools.egg-info/SOURCES.txt'
reading manifest file 'umi_tools.egg-info/SOURCES.txt'
writing manifest file 'umi_tools.egg-info/SOURCES.txt'
running build_ext
Traceback (most recent call last):
  File "setup.py", line 133, in <module>
    zip_safe=False,
  File "/ifs/apps/apps/python-2.7.10/lib/python2.7/distutils/core.py", line 151, in setup
    dist.run_commands()
  File "/ifs/apps/apps/python-2.7.10/lib/python2.7/distutils/dist.py", line 953, in run_commands
    self.run_command(cmd)
  File "/ifs/apps/apps/python-2.7.10/lib/python2.7/distutils/dist.py", line 972, in run_command
    cmd_obj.run()
  File "/ifs/apps/apps/python-2.7.10/lib/python2.7/site-packages/setuptools/command/develop.py", line 32, in run
    self.install_for_development()
  File "/ifs/apps/apps/python-2.7.10/lib/python2.7/site-packages/setuptools/command/develop.py", line 117, in install_for_development
    self.run_command('build_ext')
  File "/ifs/apps/apps/python-2.7.10/lib/python2.7/distutils/cmd.py", line 326, in run_command
    self.distribution.run_command(command)
  File "/ifs/apps/apps/python-2.7.10/lib/python2.7/distutils/dist.py", line 972, in run_command
    cmd_obj.run()
  File "/ifs/apps/apps/python-2.7.10/lib/python2.7/site-packages/Cython/Distutils/build_ext.py", line 160, in run
    if self.cython_gdb or [1 for ext in self.extensions
TypeError: 'NoneType' object is not iterable
```

TomSmithCGAT commented:
I only get this message when I run python setup.py develop rather than install. Is this the same for you?

TomSmithCGAT commented:
@IanSudbery Could you check you can run setup.py develop now?

IanSudbery commented:
setup works fine on CGAT150.

Some points I noticed:
* Lots of the help on typing `umi_tools --help` is not relevant. We do not need to look up tools by keyword!
* Utilities is listed as a umi_tools tool
* The `common options` section of help on dedup appears twice. 

More importantly, on my own machine where I use conda to manage my python environment, and had an outdated pysam, it tried to update pysam to 0.9.0, but I it won't then run.

There were some errors on pysam installation, and when I tried to run umi_tools it complained that it couldn't find pysam:

```
error: The 'pysam>=0.8.4' distribution was not found and is required by umi-tools
```

However, if I launch python and import pysam it works fine:

```
(cgat-devel-lite)mb1ims@cgat-vm:~/devel/UMI-tools$ python
Python 2.7.11 |Continuum Analytics, Inc.| (default, Dec  6 2015, 18:08:32) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://anaconda.org
>>> import pysam
>>> pysam.__version__
'0.9.0'
>>> pysam.__file__
'/shared/conda-install/envs/cgat-devel-lite/lib/python2.7/site-packages/pysam-0.0.1-py2.7-linux-x86_64.egg/pysam/__init__.pyc'
```

Clearly setuptools is returning that pysam is not install when it is. 

If I install pysam 0.8.4 via conda first, it works fine.

TomSmithCGAT commented:
Hi Ian. Just checked in on an environment using Conda to manage python packages (mac laptop) and it installed fine with pysam 0.9.0. Could you try re-installing pysam 0.9.0 and check whether it was just a problem with the installation?

TomSmithCGAT commented:
- umi_tools --help no returns a simplified help message
- Duplicated common options removed from tool specific help messages

IanSudbery commented:
Okay here is what I have learnt:

* If pysam is not installed, setup.py will not run.
* If pysam is installed, but is not a high enough version, the latest pysam will be installed, but umi_tools will not be able to find it (although the normal python interpreter can).
* If a sufficiently recent version of pysam is installed, either before, or after setup.py is run, then umi_tools will work, and will continue to work even if you uninstall pysam. 

Its all a bit confusing.

IanSudbery commented:
To reproduce on Ubuntu:

```
$conda uninstall pysam
$pip uninstall pysam

$conda install pysam=0.8.3

$python setup.py develop
```

TomSmithCGAT commented:
Yup, I can reproduce the error if I install pysam 0.8.3. At a guess I'd say this is an issue with pip install and conda? I'll have a proper look tomorrow. 

@AndreasHeger Do you know what the source of installation woes might be?

The issue with setup.py not running when pysam is not installed can be rectified easily.

TomSmithCGAT commented:
On reflection, perhaps the best solution is to make a cobra lconda package for umi-tools?

TomSmithCGAT commented:
@IanSudbery 

- Conda and Pip installations working OK.
- Integrated travis testing (Just to check installation for now, I'll add proper tests later)

Still having some issues with Bioconda package but this is a separate issue to the branch.

Are you OK for me to merge with master?
